tools/syz-crush/crush.go

```diff
@@ -24,107 +28,252 @@ import (
+
+	target, err := prog.GetTarget(cfg.TargetOS, cfg.TargetArch)
+	if err != nil {
 		log.Fatalf("%v", err)
 	}
+
 	vmPool, err := vm.Create(cfg, false)
 	if err != nil {
 		log.Fatalf("%v", err)
 	}
+
 	reporter, err := report.NewReporter(cfg)
 	if err != nil {
 		log.Fatalf("%v", err)
 	}
 
-	log.Logf(0, "booting test machines...")
+	reproduceMe := flag.Args()[0]
+	baseName := filepath.Base(reproduceMe)
+
+	runType := getType(reproduceMe)
+	if runType == CProg {
+		execprog, err := ioutil.ReadFile(reproduceMe)
+		if err != nil {
+			log.Fatalf("error reading source file from '%s'", reproduceMe)
+		}
+
+		cfg.SyzExecprogBin, err = csource.BuildNoWarn(target, execprog)
+		if err != nil {
+			log.Fatalf("failed to build source file: %v", err)
+		}
+
+		log.Logf(0, "compiled csource %v to cprog: %v", reproduceMe, cfg.SyzExecprogBin)
+	} else {
+		log.Logf(0, "reproducing from logfile: %v", reproduceMe)
+	}
+
+	restartPeriod := *flagRestartTime
+	if restartPeriod == 0 {
+		// Set default restart period to 1h
+		restartPeriod = time.Hour
+	}
+	log.Logf(0, "restartTime set to: %v", *flagRestartTime)
+
+	log.Logf(0, "booting test machines... %v", vmPool.Count())
+	runDone := make(chan *CrashReport, vmPool.Count())
 	var shutdown uint32
-	var wg sync.WaitGroup
-	wg.Add(vmPool.Count() + 1)
+	var runningWorkers uint32
+
 	for i := 0; i < vmPool.Count(); i++ {
-		i := i
-		go func() {
-			defer wg.Done()
+		atomic.AddUint32(&runningWorkers, 1)
+		go func(index int) {
 			for {
-				runInstance(cfg, reporter, vmPool, i)
-				if atomic.LoadUint32(&shutdown) != 0 {
+				runDone <- runInstance(target, cfg, reporter, vmPool, index, *flagRestartTime,
+					runType)
+				if atomic.LoadUint32(&shutdown) != 0 || !*flagInfinite {
+					atomic.AddUint32(&runningWorkers, ^uint32(0))
+
+					// If this is the last worker then we can close the channel
+					if atomic.LoadUint32(&runningWorkers) == 0 {
+						log.Logf(0, "vm-%v: closing channel", index)
+						close(runDone)
+					}
 					break
+				} else {
+					log.Logf(0, "vm-%v: restarting", index)
 				}
 			}
-		}()
+			log.Logf(0, "vm-%v: done", index)
+		}(i)
 	}
 
+	log.Logf(0, "restart/timeout set to: %v", *flagRestartTime)
 	shutdownC := make(chan struct{})
 	osutil.HandleInterrupts(shutdownC)
 	go func() {
 		<-shutdownC
-		wg.Done()
 		atomic.StoreUint32(&shutdown, 1)
 	}()
-	wg.Wait()
+
+	var count int
+	var crashes int
+	for res := range runDone {
+		count++
+		crashes += storeCrash(res, cfg, baseName)
+		log.Logf(0, "instances executed: %v", count)
+	}
+
+	log.Logf(0, "all done. reproduced %v crashes. reproduce rate %.2f%%", crashes, float64(crashes)/float64(count)*100.0)
 }
 
-func runInstance(cfg *mgrconfig.Config, reporter report.Reporter, vmPool *vm.Pool, index int) {
+func storeCrash(res *CrashReport, cfg *mgrconfig.Config, baseName string) int {
+	log.Logf(0, "storing results...")
+	if res == nil || res.Report == nil {
+		log.Logf(0, "nothing to store")
```

Comment:
This does not look useful.
It does not give any useful information to user on top of what was already logged before.
I can imagine may be you needed this for debugging, but it does not seem to be useful in the production version.